canopen/objectdictionary/__init__.py

```diff
@@ -92,8 +94,92 @@ def add_object(self, obj):
         self.indices[obj.index] = obj
         self.names[obj.name] = obj
 
-
-class Record(collections.MutableMapping):
+    def get_object(self, index, subindex):
+        if index in self.indices:
+            obj = self.indices[index]
+        elif index in self.names:
+            obj = self.names[index]
+        else:
+            # Index does not exist
+            raise OdIndexError("Index {} is not present in the "
+                               "object dictionary".format(index))
+        if not isinstance(obj, Variable):
+            # Group or array
+            if subindex not in obj:
+                # Subindex does not exist
+                raise OdSubIndexError("Object at index {} does not have a "
+                                      "subindex {}".format(index, subindex))
+            obj = obj[subindex]
+        return obj
+
+
+class ValueStore(object):
+    """Provides a common interface to set and get all values out of object
+    types that hold a collection of values."""
+    subindices = {}
+
+    @property
+    def bytes(self):
+        values = []
+        for index in self:
+            values.append(self.subindices[index].bytes)
+        return values
+
+    @bytes.setter
+    def bytes(self, values):
+        if len(values) != len(self.subindices):
+            logger.error("Length of list with new values does not match "
+                         "sub-index count")
+            return
+        for index in self:
+            self.subindices[index].bytes = values.pop(0)
+
+    @property
+    def data(self):
+        return self.bytes
+
+    @data.setter
+    def data(self, values):
+        self.bytes = values
+
+    @property
+    def raw(self):
+        """Return a list of the values of all contained sub-indices of the
+        array. The type conversion rules between CANopen and Python are
+        explained in :meth:`canopen.objectdictionary.Variable`.
+        """
+        values = []
+        for index in self:
+            values.append(self.subindices[index].raw)
+        return values
+
+    @raw.setter
+    def raw(self, values):
+        if len(values) != len(self.subindices):
+            logger.error("Length of list with new values does not match "
+                         "sub-index count")
+            return
+        for index in self:
+            self.subindices[index].raw = values.pop(0)
+
+    @property
+    def phys(self):
+        values = []
+        for index in self:
+            values.append(self.subindices[index].phys)
+        return values
+
+    @phys.setter
+    def phys(self, values):
+        if len(values) != len(self.subindices):
+            logger.error("Length of list with new values does not match "
+                         "sub-index count")
+            return
+        for index in self:
+            self.subindices[index].phys = values.pop(0)
+
+
+class Record(collections.MutableMapping, ValueStore):
```

Comment:
The ValueStore class adds a lot of attributes that doesn't make much sense to records and arrays.

Reply:
It allows to set and get all contained data at once. If this is not desired I can remove it.

Reply:
I forgot about that. Then you can keep it for now.